src/xrpld/app/tx/detail/LoanBrokerSet.cpp

```diff
@@ -0,0 +1,190 @@
+            return tecNO_PERMISSION;
+        }
+        if (account != sleBroker->at(sfOwner))
+        {
+            JLOG(ctx.j.warn()) << "Account is not the owner of the LoanBroker.";
+            return tecNO_PERMISSION;
+        }
+    }
+    else
+    {
+        auto const sleVault = ctx.view.read(keylet::vault(vaultID));
+        if (!sleVault)
+        {
+            JLOG(ctx.j.warn()) << "Vault does not exist.";
+            return tecNO_ENTRY;
+        }
+        if (account != sleVault->at(sfOwner))
+        {
+            JLOG(ctx.j.warn()) << "Account is not the owner of the Vault.";
+            return tecNO_PERMISSION;
+        }
+        if (auto const ter = canAddHolding(ctx.view, sleVault->at(sfAsset)))
+            return ter;
+    }
+    return tesSUCCESS;
+}
+
+TER
+LoanBrokerSet::doApply()
+{
+    auto const& tx = ctx_.tx;
+    auto& view = ctx_.view();
+
+    if (auto const brokerID = tx[~sfLoanBrokerID])
+    {
+        // Modify an existing LoanBroker
+        auto broker = view.peek(keylet::loanbroker(*brokerID));
+        if (!broker)
+            return tefBAD_LEDGER;  // LCOV_EXCL_LINE
+
+        if (auto const data = tx[~sfData])
+            broker->at(sfData) = *data;
+        if (auto const debtMax = tx[~sfDebtMaximum])
+            broker->at(sfDebtMaximum) = *debtMax;
+
+        view.update(broker);
+    }
+    else
+    {
+        // Create a new LoanBroker pointing back to the given Vault
+        auto const vaultID = tx[sfVaultID];
+        auto const sleVault = view.read(keylet::vault(vaultID));
+        auto const vaultPseudoID = sleVault->at(sfAccount);
+        auto const sequence = tx.getSeqValue();
+
+        auto owner = view.peek(keylet::account(account_));
+        if (!owner)
+            return tefBAD_LEDGER;  // LCOV_EXCL_LINE
+        auto broker =
+            std::make_shared<SLE>(keylet::loanbroker(account_, sequence));
+
+        if (auto const ter = dirLink(view, account_, broker))
+            return ter;
+        if (auto const ter = dirLink(view, vaultPseudoID, broker, sfVaultNode))
+            return ter;
+
+        adjustOwnerCount(view, owner, 2, j_);
+        auto const ownerCount = owner->at(sfOwnerCount);
+        if (mPriorBalance < view.fees().accountReserve(ownerCount))
+            return tecINSUFFICIENT_RESERVE;
+
+        auto maybePseudo =
+            createPseudoAccount(view, broker->key(), sfLoanBrokerID);
+        if (!maybePseudo)
+            return maybePseudo.error();
+        auto& pseudo = *maybePseudo;
+        auto pseudoId = pseudo->at(sfAccount);
+
+        if (auto ter = addEmptyHolding(
+                view, pseudoId, mPriorBalance, sleVault->at(sfAsset), j_))
+            return ter;
+
+        // Initialize data fields:
+        broker->at(sfSequence) = sequence;
+        broker->at(sfVaultID) = vaultID;
+        broker->at(sfOwner) = account_;
+        broker->at(sfAccount) = pseudoId;
+        broker->at(sfLoanSequence) = 1;
```

Comment:
In the spec we said that the loanSequence starts at zero, here it starts at one.

I updated the spec to reflect this, but perhaps this warrant a little bit of commenting:

```suggestion
        // The LoanSequence indexes loans created by this broker, starting at 1
        broker->at(sfLoanSequence) = 1;
```